### PR TITLE
Add luasocket and luasec

### PR DESCRIPTION
**Pull request by euclidianAce**
:tada: Networking! :tada: :satellite:

### CODE REVIEW

**Comment by hishamhm** [comment]
@euclidianAce Let's make the folder name match the name of the rock in luarocks.org:

* types/luasocket/socket.d.tl 
* types/luasocket/socket/ftp.d.tl 
* etc.

I plan to make this repository auto-publish these as type-declaration rocks at one point!

**Comment by hishamhm** [comment]
(added a note to that effect in the README)

**Comment by euclidianAce** [comment]
@hishamhm should I move mime/ and ltn12/ into luasocket/mime/ and luasocket/ltn12/ as well since they technically arent modules on luarocks and just come with luasocket?

**Comment by hishamhm** [comment]
yes: `types/luasocket/ltn12.d.tl` and  `types/luasocket/mime.d.tl` since it these are top-level modules in the `luasocket` package

**Comment by euclidianAce** [comment]
Should be good to go now

**Comment by hishamhm** [comment]
Looking good! But before we merge this, let me give a go at trying to fix teal-language/tl#135 later this week — then it should be possible to avoid all `global` type declarations and nest the type definitions inside the `ltn12` record.

Then it will be possible to do this from `mime.d.tl`:

```
local ltn12 = require("ltn12")
local Filter = ltn12.Filter
```

File: types/luasocket/ltn12.d.tl [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+   source = record
+      -- same situation with filter.chain and dependent types
+      cat: function<T, K>(Source<T>, Source<K>): Source<T|K>
+      cat: function<T>(Source<T>, ...: Source<T>): Source<T>
+      chain: function<T, K>(Source<T>, Filter<T, K>): Source<K>
+      empty: function(): Source<any>
+      error: function(string): Source<any>
+      file: function(FILE): Source<string>
+      file: function(FILE, string): Source<string>
+      simpilfy: function<T>(FancySource<T>): Source<T>
```
Comment (hishamhm) [inline types/luasocket/ltn12.d.tl:50]: typo:
```suggestion
      simplify: function<T>(FancySource<T>): Source<T>
```

**Comment by hishamhm** [comment]
@euclidianAce I managed to make the above stop failing, but then I realized another issue:

The `FancySink` and `FancySource` types return unions including two different functiontypes. How would a function using a FancySink or FancySource be able to tell whether they got one or the other?

Technically, I don't think this is something we can allow in Teal (and yes, the compiler is overly lenient about it right now) because given two Lua objects of type `function`, there's no way we can implement the `is` operator to destructure the union: how to tell at runtime to tell which of the functiontypes a Lua object of type `function` matches? (and I don't expect this one to be efficiently solvable, because for records we can think about tagging or using metatables)

I am going to add a restriction to at most one functiontype per union, and for FancySource and FancySink, I suggest doing just

```lua
FancySource = functiontype<T>(): T|EmptyString, string|FancySource<T>
FancySink = functiontype<T>(T|EmptyString, string): boolean, string|FancySink<T>
```

**Comment by euclidianAce** [comment]
Ah, that would be an issue. Though it seems that the Fancy versions in practice are usually converted to simple ones with `simplify`, so implementation-wise, this seems totally acceptable.

Conceptually: There probably isn't a good way to store function signatures for checking during runtime like a tagged record, would there be? I imagine it would be quite the overhead to just store them during runtime.
Is there some sort of solution at compile time of comparing signatures rather than primitive types? Or is this what already happens at some level? How that would translate to `is` working at runtime? I have no idea :shrug:

I guess I haven't really gotten a strong idea to what Teal is and isn't capable of right now.

**Comment by hishamhm** [comment]
> Conceptually: There probably isn't a good way to store function signatures for checking during runtime like a tagged record, would there be? I imagine it would be quite the overhead to just store them during runtime.

Yeah, I just mentioned about it on Gitter: for functions one _could_, say, create a weak table for the type and keep track at runtime of all functions that match the type (though generics might mess with this idea, haven't thought it through), but even then it would be a costly thing to do for little gain, and I don't expect most people to need it, and it would add a cost for everyone — I don't want to go down the path of allocating extra structures, if I can avoid it.

> Is there some sort of solution at compile time of comparing signatures rather than primitive types? Or is this what already happens at some level? 

This is already what happens at compile time, the difference is that the `is` operator is essentially the comparison-at-runtime operator.

> How that would translate to is working at runtime?

Yep, that's the issue. For tables, I'm hoping to use either something tagged records or metatables to identify different kinds of tables from each other; for primitives, we have `type()`, but all functions are `type(f) == "function"` and all functions share the same metatable.

> I guess I haven't really gotten a strong idea to what Teal is and isn't capable of right now.

We are still testing its boundaries as we develop it :) But generally the boundary is "what can translate to Lua in a straightforward manner".

> Edit: Is there a chance that functiontype could be implemented as a tagged record?

Not really, because at runtime those always map to Lua functions (and we get functions from third-party Lua libraries, so it's not like we control every object (even if we did, I don't think I'd like to wrap every function in a `__call`-able table just to make them mix together in unions)).

**Comment by hishamhm** [comment]
teal-language/tl#135 is fixed and tl 0.7.0 in released! Now it should be possible to replace all `global` type definitions with nested ones and use them between required modules :)

**Comment by hishamhm** [comment]
Fantastic!

**Pull request merged**